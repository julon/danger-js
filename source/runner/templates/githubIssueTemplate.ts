import * as v from "voca"

import { DangerResults } from "../../dsl/DangerResults"
import { Violation } from "../../dsl/Violation"

/**
 * Converts a set of violations into a HTML table
 *
 * @param {string} name User facing title of table
 * @param {string} emoji Emoji name to show next to each item
 * @param {Violation[]} violations for table
 * @returns {string} HTML
 */
function table(name: string, emoji: string, violations: Violation[]): string {
  if (violations.length === 0 || violations.every(violation => !violation.message)) {
    return ""
  }
  return `
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">${name}</th>
    </tr>
  </thead>
  <tbody>${violations
    .map((v: Violation) => {
      return `<tr>
      <td>:${emoji}:</td>
      <td>

  ${v.message}
  </td>
    </tr>
  `
    })
    .join("\n")}</tbody>
</table>
`
}

function getSummary(label: string, violations: Violation[]): string {
  return violations
    .map(x => v.truncate(x.message, 20))
    .reduce(
      (acc, value, idx) => `${acc} ${value}${idx === violations.length - 1 ? "" : ","}`,
      `${violations.length} ${label}: `
    )
}

function buildSummaryMessage(dangerID: string, results: DangerResults): string {
  const { fails, warnings, messages, markdowns } = results
  const summary = `  ${getSummary("failure", fails)}
  ${getSummary("warning", warnings)}
  ${messages.length > 0 ? `${messages.length} messages` : ""}
  ${markdowns.length > 0 ? `${markdowns.length} markdown notices` : ""}
  ${dangerIDToString(dangerID)}`
  return summary
}

export const dangerIDToString = (id: string) => `DangerID: danger-id-${id};`

/**
 * Postfix signature to be attached comment generated / updated by danger.
 */
export const dangerSignaturePostfix = `Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>`

/**
 * A template function for creating a GitHub issue comment from Danger Results
 * @param {string} dangerID A string that represents a unique build
 * @param {DangerResults} results Data to work with
 * @returns {string} HTML
 */
export function template(dangerID: string, results: DangerResults): string {
  return `
<!--
${buildSummaryMessage(dangerID, results)}
-->
${table("Fails", "no_entry_sign", results.fails)}
${table("Warnings", "warning", results.warnings)}
${table("Messages", "book", results.messages)}
${results.markdowns.join("\n\n")}
<p align="right">
  ${dangerSignaturePostfix}
</p>
`
}
