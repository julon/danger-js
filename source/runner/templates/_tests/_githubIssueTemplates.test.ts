import {
  emptyResults,
  failsResultsWithoutMessages,
  warnResults,
  failsResults,
  summaryResults,
} from "../../_tests/fixtures/ExampleDangerResults"
import { template as githubResultsTemplate } from "../../templates/githubIssueTemplate"

describe("generating messages", () => {
  it("shows no tables for empty results", () => {
    const issues = githubResultsTemplate("blankID", emptyResults)
    expect(issues).not.toContain("Fails")
    expect(issues).not.toContain("Warnings")
    expect(issues).not.toContain("Messages")
  })

  it("shows no tables for results without messages", () => {
    const issues = githubResultsTemplate("blankID", failsResultsWithoutMessages)
    expect(issues).not.toContain("Fails")
    expect(issues).not.toContain("Warnings")
    expect(issues).not.toContain("Messages")
  })

  it("Shows the failing messages in a table", () => {
    const issues = githubResultsTemplate("blankID", failsResults)
    expect(issues).toContain("Fails")
    expect(issues).not.toContain("Warnings")
  })

  it("Shows the warning messages in a table", () => {
    const issues = githubResultsTemplate("blankID", warnResults)
    expect(issues).toContain("Warnings")
    expect(issues).not.toContain("Fails")
  })

  it("does not break commonmark rules around line breaks", () => {
    const issues = githubResultsTemplate("blankID", warnResults)
    expect(issues).not.toMatch(/(\r?\n){2}[ \t]+</)
  })

  it("Should include summary on top of message", () => {
    const issues = githubResultsTemplate("blankID", summaryResults)
    const expected = `
<!--
  1 failure:  Failing message F...
  1 warning:  Warning message W...
  1 messages
  1 markdown notices
  DangerID: danger-id-blankID;
-->`

    expect(issues).toContain(expected)
  })

  it("leaves space between <td>s to allow GitHub to render message content as markdown", () => {
    const issues = githubResultsTemplate("example-id", {
      fails: [{ message: "**Failure:** Something failed!" }],
      warnings: [{ message: "_Maybe you meant to run `yarn install`?_" }],
      messages: [{ message: "```ts\nfunction add(a: number, b: number): number {\n  return a + b\n}\n```" }],
      markdowns: ["List of things:\n\n* one\n* two\n* three\n"],
    })

    expect(issues).toMatchSnapshot()
  })
})
