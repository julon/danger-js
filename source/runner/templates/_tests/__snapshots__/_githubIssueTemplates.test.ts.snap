// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generating messages leaves space between <td>s to allow GitHub to render message content as markdown 1`] = `
"
<!--
  1 failure:  **Failure:** Some...
  1 warning:  _Maybe you meant ...
  1 messages
  1 markdown notices
  DangerID: danger-id-example-id;
-->

<table>
  <thead>
    <tr>
      <th width=\\"50\\"></th>
      <th width=\\"100%\\" data-danger-table=\\"true\\">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Failure:** Something failed!
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width=\\"50\\"></th>
      <th width=\\"100%\\" data-danger-table=\\"true\\">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  _Maybe you meant to run \`yarn install\`?_
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width=\\"50\\"></th>
      <th width=\\"100%\\" data-danger-table=\\"true\\">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  \`\`\`ts
function add(a: number, b: number): number {
  return a + b
}
\`\`\`
  </td>
    </tr>
  </tbody>
</table>

List of things:

* one
* two
* three

<p align=\\"right\\">
  Generated by :no_entry_sign: <a href=\\"http://github.com/danger/danger-js/\\">dangerJS</a>
</p>
"
`;
